diff --git a/Makefile b/Makefile
index c244a02..96d353d 100644
--- a/Makefile
+++ b/Makefile
@@ -185,9 +185,6 @@ SUBARCH := $(shell uname -m | sed -e s/i.86/i386/ -e s/sun4u/sparc64/ \
 ARCH		?= $(SUBARCH)
 CROSS_COMPILE	?=
 
-# Architecture as present in compile.h
-UTS_MACHINE := $(ARCH)
-
 KCONFIG_CONFIG	?= .config
 
 # SHELL used by kbuild
@@ -321,6 +318,11 @@ AFLAGS          := -D__ASSEMBLY__
 KERNELRELEASE = $(shell cat include/config/kernel.release 2> /dev/null)
 KERNELVERSION = $(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
 
+-include .kernelvariables
+
+# Architecture as present in compile.h
+UTS_MACHINE := $(ARCH)
+
 export VERSION PATCHLEVEL SUBLEVEL KERNELRELEASE KERNELVERSION
 export ARCH CONFIG_SHELL HOSTCC HOSTCFLAGS CROSS_COMPILE AS LD CC
 export CPP AR NM STRIP OBJCOPY OBJDUMP MAKE AWK GENKSYMS PERL UTS_MACHINE
